File: tier4_expert/monotonic_stack/solutions/p853_car_fleet.py
"""P853: Car Fleet (Medium)

TODO: Implement solve() below.
Hint: Sort by position desc, calculate time to target. A car forms a new fleet only if it arrives before the current fleet.
"""
import sys
sys.path.insert(0, ".")
from src.utils import Problem, TestCase


class Solution(Problem):
    name = "853. Car Fleet"
    test_cases = [
        TestCase(input=(12, [10,8,0,5,3], [2,4,1,1,3]), expected=3, label="example 1"),
        TestCase(input=(10, [3], [3]), expected=1, label="example 2"),
        TestCase(input=(100, [0,2,4], [4,2,1]), expected=1, label="all merge"),
    ]

    def solve(self, target: int, position: list[int], speed: list[int]) -> int:
        # Pair position with speed, sort by position descending
        cars = sorted(zip(position, speed), key=lambda x: -x[0])
        fleets = 0
        slowest_time = -1.0

        for pos, spd in cars:
            time = (target - pos) / spd
            if time > slowest_time:
                fleets += 1
                slowest_time = time

        return fleets


if __name__ == "__main__":
    Solution().run()

File: tier4_expert/monotonic_stack/solutions/p853_car_fleet.mjs
/**
 * P853: Car Fleet (Medium)
 * https://leetcode.com/problems/car-fleet/
 * Topics: Array, Stack, Sorting, Monotonic Stack
 *
 * There are n cars at given miles away from the starting mile 0, traveling to reach the mile target.
 * You are given two integer arrays position and speed, both of length n, where position[i] is the starting mile of the ith car and speed[i] is the speed of the ith car in miles per hour.
 * A car cannot pass another car, but it can catch up and then travel next to it at the speed of the slower car.
 * Return the number of car fleets that will arrive at the destination.
 *
 * Example 1:
 *     Input: target = 12, position = [10,8,0,5,3], speed = [2,4,1,1,3]
 *     Output: 3
 *
 * Example 2:
 *     Input: target = 10, position = [3], speed = [3]
 *     Output: 1
 *
 * Example 3:
 *     Input: target = 100, position = [0,2,4], speed = [4,2,1]
 *     Output: 1
 *
 * Hint: Sort by position desc, calculate time to target. A car forms a new fleet only if it arrives before the current fleet.
 */

import { readLine, readInts, readInt, writeInt, writeInts, writeString, writeBool } from '../../../wasm_libs/js/io.mjs';

function solve(target, position, speed) {
  const cars = position.map((p, i) => [p, speed[i]]);
  cars.sort((a, b) => b[0] - a[0]);
  let fleets = 0;
  let slowestTime = -1;
  for (const [pos, spd] of cars) {
    const time = (target - pos) / spd;
    if (time > slowestTime) {
      fleets++;
      slowestTime = time;
    }
  }
  return fleets;
}

const target = readInt();
const position = readInts();
const speed = readInts();
writeInt(solve(target, position, speed));
